string: Always compile SVE routines

xboxfanj opened the pull request:
This is a partial revert of b7e368fb.
If SVE assembly is guarded by __ARM_FEATURE_SVE,
it cannot build when SVE is not enabled by the
build system. This is ok on AOR, but because
Android (bionic) uses ifuncs to select the
appropriate assembly at runtime, these
need to compile regardless of if the target
actually supports the instructions.

xboxfanj commented:
@Wilco1 should I just apply the change to memcpy in that case since it has been optimized?

Regarding compilers, we could also check if the clang version or GCC version is high enough if that would be better.

xboxfanj commented:
Thanks @Wilco1, I just changed the commit to only affect memcpy/memmove and only build on supported compilers.  Please take a look when you can.

fyi @enh-google

nsz-arm commented on string/test/memcpy.c:
```diff
@@ -28,7 +28,7 @@ static const struct fun
   F(__memcpy_aarch64_simd, 1)
 # endif
-# if __ARM_FEATURE_SVE
+# ifdef HAVE_SVE
   F(__memcpy_aarch64_sve, 1)
```
to run the test it is not enough that the compiler supports SVE, the hw has to support it too.

we don't have runtime check for SVE so i'd leave the original ifdef check here (or add the runtime check: for each function specify a hwcap mask and if not all bits are set then skip that function).

nsz-arm commented on string/test/memmove.c:
```diff
@@ -28,7 +28,7 @@ static const struct fun
   F(__memmove_aarch64_simd, 1)
 # endif
-# if __ARM_FEATURE_SVE
+# ifdef HAVE_SVE
   F(__memmove_aarch64_sve, 1)
```
likewise

nsz-arm reviewed (changes requested):
this is ok with the comments fixed.

for testing and benchmarking runtime checks are needed (getauxval + HWCAP_SVE check) to enable the sve routines. but i think this should be a separate commit.

Wilco1 commented on string/include/stringlib.h:
```diff
@@ -37,9 +37,11 @@ char *__strrchr_aarch64_mte (const char *, int);
 void *__memmove_aarch64_simd (void *, const void *, size_t);
 #endif
-# if __ARM_FEATURE_SVE
+#ifdef HAVE_SVE
```
Do we need these defines for the header? It seems unnecessary (and out of sync if we use __ARM_FEATURE_SVE in some places and HAVE_SVE in others...).

nsz-arm replied:
it is not needed but does not hurt either.. except the HAVE_SVE is only defined for asm code so it should not be used in c code at all.  for now i think the code should be left alone and the header fixed up in a separate commit that moves to runtime feature check.

xboxfanj replied:
Done.

xboxfanj commented:
Thanks @nsz-arm, I addressed your comments in the latest push.

Should I also match the arch flag from https://github.com/ARM-software/optimized-routines/blob/master/config.mk.dist#L68?

nsz-arm commented:
i think armv8-a+sve arch is fine.

nsz-arm commented:
thanks. committed with reformatted commit message at https://github.com/ARM-software/optimized-routines/commit/cd28f3c4253257027f9834eb7b17db047346a267
